iocage_lib/ioc_common.py

```diff
@@ -740,30 +740,45 @@ def generate_devfs_ruleset(conf, paths=None, includes=None, callback=None,
     Will add a per jail devfs ruleset with the specified rules,
     specifying defaults that equal devfs_ruleset 4
     """
-    ruleset = conf['devfs_ruleset']
+    configured_ruleset = conf['devfs_ruleset']
     devfs_includes = []
     devfs_rulesets = su.run(
         ['devfs', 'rule', 'showsets'],
         stdout=su.PIPE, universal_newlines=True
     )
     ruleset_list = [int(i) for i in devfs_rulesets.stdout.splitlines()]
 
-    if ruleset != '4':
-        if int(ruleset) in ruleset_list:
-            return str(ruleset)
+    # 4 is a magic number for default and doesn't refer to
+    # the actual ruleset 4 in devfs.rules(!)
+    default_ruleset = 4
+    min_dynamic_ruleset = 1000  # leave lower numbers to sysadmin
 
-        logit({
-            "level": "INFO",
-            "message": f'* Ruleset {ruleset} does not exist, using defaults'
-        },
-            _callback=callback,
-            silent=silent)
-
-    ruleset = 5  # 0-4 is always reserved
+    ruleset = min_dynamic_ruleset
     while ruleset in ruleset_list:
         ruleset += 1
     ruleset = str(ruleset)
 
+    # Custom devfs_ruleset configured, clone to dynamic ruleset
+    if int(configured_ruleset) != default_ruleset:
+        if int(configured_ruleset) not in ruleset_list:
+            logit(
+                {
+                    "level": "EXCEPTION",
+                    "message": f"devfs_ruleset {ruleset} doesn't exist"
```

Comment:
@grembo can we please catch the exceptions which this would raise in `ioc_start.py` and raise an appropriate error message saying that `Failed to start {jail_name}: the current one we have right now here`. Thoughts ?

Reply:
Hi @sonicaj,

Thanks for looking into this. I changed the code now, so it shows a proper error message in case a devfs_ruleset isn't configured.

While doing so I found a couple of more issues that I addressed, please see the commit message of the new patch. This makes the patch a bit bigger, but also *a lot* more useful. 

Let me know if you have any questions.

See below on what startup messages look like now.

devfs_ruleset non-existent:
```
# iocage start myjail
* Starting myjail
myjail devfs_ruleset 777 doesnt_exit! - Not starting jail
```

devfs_ruleset existent:
```
# iocage start myjail
* Starting myjail
  + Started OK
  + Cloning devfs_ruleset 501
  + Using devfs_ruleset: 1006
  + Using IP options: ip4.addr=lo1|10.0.111.21/32...
  + Starting services OK
  + Executing poststart OK
```

devfs_ruleset default/unconfigured:
```
# iocage start myjail
* Starting myjail
  + Started OK
  + Generating dynamic devfs_ruleset
  + Using devfs_ruleset: 1006
  + Using IP options: ip4.addr=lo1|10.0.111.21/32...
  + Starting services OK
  + Executing poststart OK
```